openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/workspace/Market_with_Director.ipynb

```diff
@@ -0,0 +1,2183 @@
+    "        \"\"\"Get item from dataset.\"\"\"\n",
+    "        img, (pid, camid) = self.dataset[index]\n",
+    "        if self.transform is not None:\n",
+    "            img = self.transform(img)\n",
+    "        return img, (pid, camid)\n",
+    "\n",
+    "\n",
+    "class MarketFLDataloader(DataInterface):\n",
+    "    \"\"\"Market Dataset.\"\"\"\n",
+    "\n",
+    "    def __init__(self, **kwargs):\n",
+    "        super().__init__(**kwargs)\n",
+    "\n",
+    "        # Prepare transforms\n",
+    "        self.transform_train = Compose([\n",
+    "            T.ResizeRandomCropping(256, 128, p=0.5),\n",
+    "            RandomHorizontalFlip(),\n",
+    "            ToTensor(),\n",
+    "            Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),\n",
+    "            T.RandomErasing(probability=0.5)\n",
+    "        ])\n",
+    "        self.transform_test = Compose([\n",
+    "            Resize((265, 128)),\n",
+    "            ToTensor(),\n",
+    "            Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),\n",
+    "        ])\n",
+    "\n",
+    "    @property\n",
+    "    def shard_descriptor(self):\n",
+    "        return self._shard_descriptor\n",
+    "\n",
+    "    @shard_descriptor.setter\n",
+    "    def shard_descriptor(self, shard_descriptor):\n",
+    "        \"\"\"\n",
+    "        Describe per-collaborator procedures or sharding.\n",
+    "\n",
+    "        This method will be called during a collaborator initialization.\n",
+    "        Local shard_descriptor will be set by Envoy.\n",
+    "        \"\"\"\n",
+    "        self._shard_descriptor = shard_descriptor\n",
+    "\n",
+    "    def get_train_loader(self, **kwargs):\n",
+    "        \"\"\"\n",
+    "        Output of this method will be provided to tasks with optimizer in contract.\n",
+    "        \"\"\"\n",
+    "        if self.kwargs['train_bs']:\n",
+    "            batch_size = self.kwargs['train_bs']\n",
+    "        else:\n",
+    "            batch_size = 64\n",
+    "\n",
+    "        self.shard_descriptor.set_mode('train')\n",
+    "\n",
+    "        return DataLoader(\n",
+    "            # ImageDataset make transform\n",
+    "            ImageDataset(self.shard_descriptor, transform=self.transform_train),\n",
+    "            sampler=RandomIdentitySampler(self.shard_descriptor, num_instances=4),\n",
+    "            batch_size=batch_size, num_workers=4, pin_memory=True, drop_last=True\n",
+    "        )\n",
+    "\n",
+    "    def get_valid_loader(self, **kwargs):\n",
+    "        \"\"\"\n",
+    "        Output of this method will be provided to tasks without optimizer in contract.\n",
+    "        \"\"\"\n",
+    "        if self.kwargs['valid_bs']:\n",
+    "            batch_size = self.kwargs['valid_bs']\n",
+    "        else:\n",
+    "            batch_size = 512\n",
+    "        \n",
+    "        self.shard_descriptor.set_mode('query')\n",
```

Comment:
We probably should pass `shard_descriptor` to DataLoader by deep copying it. 
When the Pytorch DataLoader will try to access shard_descriptor the mode may be already changed

Reply:
added shard descriptor's copies to gallery and query loaders